src/mqttgw_dyn_srv.erl

```diff
@@ -0,0 +1,121 @@
+                [Data]),
+            {error, #{reason_code => impl_specific_error}};
+        %% if the subject is subscribed on multiple nodes something went wrong
+        %% since we disallow same agent ids
+        NodeSubs when length(NodeSubs) > 1 ->
+            error_logger:error_msg(
+                "Error on publish: subject was subscribed to multiple nodes simultaneously, data: ~p",
+                [Data]),
+            {error, #{reason_code => impl_specific_error}};
+        %% happy case when subject is on the same node as we are
+        [{Node, _, _Subs}] when Node == CurrentNode ->
+            create_sub(Subject, Data),
+            {CorrData, RespTopic, SenderConn, SenderId, UniqueId, SessionPairId, Time} = DynsubRespData,
+
+            %% Send a response to the application.
+            mqttgw:send_dynsub_response(
+                CorrData, RespTopic, SenderConn, SenderId,
+                UniqueId, SessionPairId, Time),
+            ok;
+        %% Subject is on a different node, we need to wait for metadata to propagate to that node.
+        [{Node, _, _Subs}] ->
+            Topic = authz_subscription_topic(Data),
+            start_dynsub(Node, Subject, Topic, Data, DynsubRespData),
+            ok
+    end.
+
+start_dynsub(Node, Subject, Topic, Data, DynsubRespData) ->
+    gen_server:cast({global, {?MODULE, Node}}, {start_dynsub, Subject, Topic, Data, DynsubRespData, node()}).
+
+register_dynsub_in_progress(Node, Subject, Topic, DynsubRespData) ->
+    gen_server:cast({global, {?MODULE, Node}}, {register_dynsub_in_progress, Subject, Topic, DynsubRespData}).
+
+
+-spec authz_subscription_topic(mqttgw_dynsub:data()) -> mqttgw:topic().
+authz_subscription_topic(Data) ->
+    #{app := App,
+      object := Object,
+      version := Version} = Data,
+    [<<"apps">>, App, <<"api">>, Version | Object].
+
+-spec create_sub(mqttgw_dynsub:subject(), mqttgw_dynsub:data()) -> ok.
+create_sub(Subject, Data) ->
+    QoS = 1,
+    Topic = authz_subscription_topic(Data),
+    mqttgw_broker:subscribe(Subject, [{Topic, QoS}]),
+
+    error_logger:info_msg(
+        "Dynamic subscription: ~p has been created "
+        "for the subject = '~s'",
+        [Data, Subject]),
+
+    ok.
+
+init([]) ->
+    EHandler = vmq_subscriber_db:subscribe_db_events(),
+    State = #{event_handler => EHandler, subs_queue => #{}},
+    {ok, State}.
+
+handle_cast({start_dynsub, Subject, Topic, Data, DynsubRespData, From}, State) ->
+    register_dynsub_in_progress(From, Subject, Topic, DynsubRespData),
+    create_sub(Subject, Data),
+    {noreply, State};
+handle_cast({register_dynsub_in_progress, Subject, Topic, DynsubRespData}, #{subs_queue := SubsQueue} = State) ->
+    SubscriberQ = maps:get(Subject, SubsQueue, #{}),
+    NewSubscriberQ = maps:put(Topic, DynsubRespData, SubscriberQ),
+    NewSubsQueue = maps:put(Subject, NewSubscriberQ, SubsQueue),
+    NewState = maps:put(subs_queue, NewSubsQueue, State),
+    {noreply, NewState};
```

Comment:
It may be a bit clearer to update the map in the following way:
```erlang
handle_cast(...) ->
    SubscriberQ = maps:get(Subject, SubsQueue, #{})
    NewState = State#{
        subs_queue => SubsQueue#{
            Subject => SubscriberQ#{
                Topic => DynsubRespData}}}
    {noreply, NewState};
```